### PR TITLE
adding base_backup

### PR DESCRIPTION
**Pull request by sebasmannem**
Started working on adding BASE_BACKUP functionality.
More info, read https://www.postgresql.org/docs/current/protocol-replication.html

### CODE REVIEW

**Comment by sebasmannem** [comment]
@jackc could you have a look at this PR?
Let me know if it needs anything, or raises questions.
Thanks :)

**Comment by jackc** [comment]
I'm not very familiar with the base backup protocol so I can only really review it at a surface level. I'll add comments at the line level for them.

The long test is unfortunate, but I think its better than not being tested at all. However, it can be too long -- it appears to depend on the size of the database cluster.  I do most of my testing locally (no Docker or containers) so it tries to backup my local main database cluster that is quite large. It fails with a timeout. I suppose it really needs a fresh database cluster. It might be nice to note this in the README test setup instructions.

File: pglogrepl.go [hunk 1]
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
...
 
+type BaseBackupOptions struct {
+	// Request information required to generate a progress report, but might as such have a negative impact on the performance.
+	Progress          bool
+	// Sets the label of the backup. If none is specified, a backup label of 'wal-g' will be used.
+	Label             string
+	// Request a fast checkpoint.
+	Fast              bool
+	// Include the necessary WAL segments in the backup. This will include all the files between start and stop backup in the pg_wal directory of the base directory tar file.
+	Wal               bool
```
Comment (jackc) [inline pglogrepl.go:294]: I think that this should be `WAL` to conform to Go naming conventions for acronyms.

File: pglogrepl.go [hunk 2]
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
...
+	// By default, the backup will wait until the last required WAL segment has been archived, or emit a warning if log archiving is not enabled.
+	// Specifying NOWAIT disables both the waiting and the warning, leaving the client responsible for ensuring the required log is available.
+	NoWait            bool
+	// Limit (throttle) the maximum amount of data transferred from server to client per unit of time (kb/s).
+	MaxRate           int
```
Comment (jackc) [inline pglogrepl.go:299]: I think this is a known size field. I would think this should be `int32` or `int64`.

File: pglogrepl.go [hunk 3]
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
...
+	}
+	if bbo.NoVerifyChecksums {
+		parts = append(parts, "NOVERIFY_CHECKSUMS")
+	}
+	return strings.Join(parts, " ")
+}
+
+// BaseBackupTablespace represents a tablespace in the backup
+type BaseBackupTablespace struct {
+	Oid int
```
Comment (jackc) [inline pglogrepl.go:338]: `Oid` -> `OID`. OIDs are `uint32` in PostgreSQL.

File: pglogrepl.go [hunk 4]
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
...
+	Location string
+	Size int8
+}
+
+// BaseBackupResult will hold the return values  of the BaseBackup command
+type BaseBackupResult struct {
+	Lsn LSN
```
Comment (jackc) [inline pglogrepl.go:345]: `Lsn` -> `LSN`

File: pglogrepl.go [hunk 5]
```diff
@@ -283,6 +283,230 @@ func StartReplication(ctx context.Context, conn *pgconn.PgConn, slotName string,
...
+	// Limit (throttle) the maximum amount of data transferred from server to client per unit of time (kb/s).
+	MaxRate           int
+	// Include information about symbolic links present in the directory pg_tblspc in a file named tablespace_map.
+	TablespaceMap     bool
+	// Disable checksums being verified during a base backup.
+	// Note that NoVerifyChecksums=true is only supported since PG11
+	NoVerifyChecksums bool
+}
+
+func (bbo BaseBackupOptions) Sql() string {
```
Comment (jackc) [inline pglogrepl.go:307]: `Sql` -> `SQL`

File: pglogrepl.go (hunk 1 above)
Comment (sebasmannem) [reply pglogrepl.go:294]: Agree. CHanged.

File: pglogrepl.go (hunk 2 above)
Comment (sebasmannem) [reply pglogrepl.go:299]: int32 is fine.

File: pglogrepl.go (hunk 3 above)
Comment (sebasmannem) [reply pglogrepl.go:338]: Changed

File: pglogrepl.go (hunk 4 above)
Comment (sebasmannem) [reply pglogrepl.go:345]: Changed

File: pglogrepl.go (hunk 5 above)
Comment (sebasmannem) [reply pglogrepl.go:307]: Yeah, but it could be private too, so I changed it to sql.

**Comment by sebasmannem** [comment]
> The long test is unfortunate, but I think its better than not being tested at all....
I have added an option to disable the test if you want to.
> It fails with a timeout. 
I removed the timeouts.
> It might be nice to note this in the README test setup instructions.
I documented the option to disable it in the README file.

I also noticed some minor issues in tests (with tablespaces, and with higher incarnation), so I fixed them too.

**Comment by jackc** [comment]
Looks pretty good. Just two more tweaks.

1. OID should be `uint32` instead of `int32`. Unlikely to cause any problems, but still better to be entirely correct.
2. There is a binary in one of the commits. If you can conveniently rebase / rewrite history to remove it and avoid bloating the repo that would be great. Otherwise, I can squash merge to remove it.

**Comment by sebasmannem** [comment]
Ouch. Binary was in last commit. I removed it.
And changed int32 to uint32.

**Comment by jackc** [comment]
Thanks!

**Pull request merged**